jni/src/faiss_index_service.cpp

```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// The OpenSearch Contributors require contributions made to
+// this file be licensed under the Apache-2.0 license or a
+// compatible open source license.
+//
+// Modifications Copyright OpenSearch Contributors. See
+// GitHub history for details.
+
+#include "faiss_index_service.h"
+#include "faiss_methods.h"
+#include "faiss/index_factory.h"
+#include "faiss/Index.h"
+#include "faiss/IndexBinary.h"
+#include "faiss/IndexHNSW.h"
+#include "faiss/IndexBinaryHNSW.h"
+#include "faiss/IndexIVFFlat.h"
+#include "faiss/IndexBinaryIVF.h"
+#include "faiss/IndexIDMap.h"
+#include "faiss/index_io.h"
+#include <algorithm>
+#include <string>
+#include <vector>
+#include <memory>
+#include <type_traits>
+
+namespace knn_jni {
+namespace faiss_wrapper {
+
+template<typename INDEX, typename IVF, typename HNSW>
+void SetExtraParameters(knn_jni::JNIUtilInterface * jniUtil, JNIEnv *env,
```

Comment:
really good abstraction.. Just by looking at it feels overkill, but when I looked at the reusability of the code looks absolutely awesome. But not sure if this is the best way to achieve this.

Reply:
Yeah. It was my best effort. There could be better a way though.

Reply:
I think I am good on this for now. Till someone contribute a better way that this.